### DISCUSSION TITLE
Setting User/Group failed

### DISCUSSION THREAD

**Discussion by alsimfer** [rustic-rs/rustic#503] [Things are not working as expected - is this a bug?] [answered]
Hi,

TLDR:
Can I skip setting of original user:group ids during `rustic restore`?

Longer read:
We use restic for backups in K8S. Other than for complete restores, we want also to use those backups to keep QA envs updated with the data created in Prod. So we basically want delta migrations which restic doesnt support. Special tnx for this feature in rustic! Now the issue: during restore the rustic seem to try setting original user:group permissions on all files. It takes long, is not needed and only produces errors, because the target containers are running under different users

**Comment by aawsome** [comment]
You can tell rustic to set uid/gid instead of user/group. See the option `--numeric-id`.
Also, I can easily add an option to not set these at all, please open an issue for this if you need it.

**Reply by LeonardoMantovani** [reply]
I’m using rustic and r clone to backup the volumes of a nextcloud docker container, where some files have the ownership set to user 33 and group 33.
But if I try to restore the snapshot with `--numeric-id` it returns me a list of [WARN] saying *setting UID/GID failed.*

I guess it fails because the uid and gid used do not exist in my host (they only make sense inside the container), but it still shouldn’t be a problem because if I try to do `chown 33:33` on one of the restored files the system lets me do so whiteout problems. **So is there a flag I should pass to rustic to make it not check the existence of the uid and gid and just restore them?** Or is there a different way I can achieve the same thing without having to move back to restic?

**Reply by aawsome** [reply]
@LeonardoMantovani Thanks for asking! Actually the `--numeric-id` option should do exactly what `chown` is doing...
Can you have a look at the trees in you repository which contain the file/dir where this is failing? This can be done using `rustic cat tree <SNAP>:<PATH>?`, e.g. `rustic cat tree a5c7a:/my/path`.  The output is JSON, so you might want to pipe through `jq`.
This should tell you which UID/GID rustic wants to set. If it really `33:33` we should dig deeper...

(Unfortunately there is so far no `rustic ls --numeric-ic` or something like this)

**Reply by LeonardoMantovani** [reply]
Thank you for your fast response. I'll check this out as soon as possible and share the output here.

**Reply by LeonardoMantovani** [reply]
Ok I tried out running the command `rustic -r rclone:storj:mybackup/nextcloud cat tree latest:nextcloud/data_v/config` and this was the output:

```json
{
  "nodes": [
    {
      "name": ".htaccess",
      "type": "file",
      "mode": 420,
      "mtime": "2023-12-27T10:48:10.423799375+01:00",
      "atime": "2023-12-27T10:48:10.423799375+01:00",
      "ctime": "2023-12-27T19:05:28.408866058+01:00",
      "uid": 33,
      "gid": 33,
      "inode": 9583918,
      "device_id": 66306,
      "size": 495,
      "links": 1,
      "content": [
        "a7f1448e2abdf7373ef6c39ff501e4b0d7341314512668bd1def0b13ac233508"
      ]
    },
    {
      "name": "apache-pretty-urls.config.php",
      "type": "file",
      "mode": 420,
      "mtime": "2023-12-27T10:48:10.423799375+01:00",
      "atime": "2023-12-27T10:48:10.423799375+01:00",
      "ctime": "2023-12-27T19:05:28.408866058+01:00",
      "uid": 33,
      "gid": 33,
      "inode": 9583920,
      "device_id": 66306,
      "size": 60,
      "links": 1,
      "content": [
        "9d881cf59f450d5434ac146091b77e9544abf868d8b6c704482c54b9ef0ccf84"
      ]
    },
    {
      "name": "apcu.config.php",
      "type": "file",
      "mode": 420,
      "mtime": "2023-12-27T10:48:10.423799375+01:00",
      "atime": "2023-12-27T10:48:10.423799375+01:00",
      "ctime": "2023-12-27T19:05:28.408866058+01:00",
      "uid": 33,
      "gid": 33,
      "inode": 9583921,
      "device_id": 66306,
      "size": 70,
      "links": 1,
      "content": [
        "d66658b683c7e96513a8fa06aab1dc7df7c69df969d1103db78fdff493b8cb59"
      ]
    },
    {
      "name": "apps.config.php",
      "type": "file",
      "mode": 420,
      "mtime": "2023-12-27T10:48:10.423799375+01:00",
      "atime": "2023-12-27T10:48:10.423799375+01:00",
      "ctime": "2023-12-27T19:05:28.408866058+01:00",
      "uid": 33,
      "gid": 33,
      "inode": 9583922,
      "device_id": 66306,
      "size": 377,
      "links": 1,
      "content": [
        "92817524be6545c000b00ad27fc7cfc57547d3ad2da4de11c56c400b63fb67b8"
      ]
    },
    {
      "name": "config.php",
      "type": "file",
      "mode": 416,
      "mtime": "2024-01-26T03:30:09.412236108+01:00",
      "atime": "2024-01-26T03:30:09.412236108+01:00",
      "ctime": "2024-01-26T03:30:09.412236108+01:00",
      "uid": 33,
      "gid": 33,
      "inode": 9583954,
      "device_id": 66306,
      "size": 1680,
      "links": 1,
      "content": [
        "bdaf68d81bfea0b54d157e33da6bc26787787730df5f01b088d8b0dbd6229d9c"
      ]
    },
    {
      "name": "config.sample.php",
      "type": "file",
      "mode": 420,
      "mtime": "2023-12-27T10:48:10.424799432+01:00",
      "atime": "2023-12-27T10:48:10.424799432+01:00",
      "ctime": "2023-12-27T19:05:28.408866058+01:00",
      "uid": 33,
      "gid": 33,
      "inode": 9583924,
      "device_id": 66306,
      "size": 76941,
      "links": 1,
      "content": [
        "2bc9d3a7f08ca0f8f7ecfedbea8bce0e41503ddd1c95ec7915685b34ab97849e"
      ]
    },
    {
      "name": "redis.config.php",
      "type": "file",
      "mode": 420,
      "mtime": "2023-12-27T10:48:10.424799432+01:00",
      "atime": "2023-12-27T10:48:10.424799432+01:00",
      "ctime": "2023-12-27T19:05:28.408866058+01:00",
      "uid": 33,
      "gid": 33,
      "inode": 9583925,
      "device_id": 66306,
      "size": 484,
      "links": 1,
      "content": [
        "737eaaac44e7d8bbd633214949969dc9b1e2daaff44be029cb9d990c16b88d37"
      ]
    },
    {
      "name": "reverse-proxy.config.php",
      "type": "file",
      "mode": 420,
      "mtime": "2023-12-27T10:48:10.424799432+01:00",
      "atime": "2023-12-27T10:48:10.424799432+01:00",
      "ctime": "2023-12-27T19:05:28.408866058+01:00",
      "uid": 33,
      "gid": 33,
      "inode": 9583926,
      "device_id": 66306,
      "size": 798,
      "links": 1,
      "content": [
        "a6e6f1939619cef068377bdd2ab65c36c3e34c77f0ac6d9b796941159fad2712"
      ]
    },
    {
      "name": "s3.config.php",
      "type": "file",
      "mode": 420,
      "mtime": "2023-12-27T10:48:10.424799432+01:00",
      "atime": "2023-12-27T10:48:10.424799432+01:00",
      "ctime": "2023-12-27T19:05:28.408866058+01:00",
      "uid": 33,
      "gid": 33,
      "inode": 9583927,
      "device_id": 66306,
      "size": 2030,
      "links": 1,
      "content": [
        "30e79be54211bff779de6f3718bd9ae99e9b1d5996b20d3e0cf4d9e4bece1676"
      ]
    },
    {
      "name": "smtp.config.php",
      "type": "file",
      "mode": 420,
      "mtime": "2023-12-27T10:48:10.424799432+01:00",
      "atime": "2023-12-27T10:48:10.424799432+01:00",
      "ctime": "2023-12-27T19:05:28.408866058+01:00",
      "uid": 33,
      "gid": 33,
      "inode": 9583928,
      "device_id": 66306,
      "size": 1025,
      "links": 1,
      "content": [
        "6c06309c7d3923a4b53ec59236d684497310bf386600092ef00d11c9d1d542f6"
      ]
    },
    {
      "name": "swift.config.php",
      "type": "file",
      "mode": 420,
      "mtime": "2023-12-27T10:48:10.424799432+01:00",
      "atime": "2023-12-27T10:48:10.424799432+01:00",
      "ctime": "2023-12-27T19:05:28.408866058+01:00",
      "uid": 33,
      "gid": 33,
      "inode": 9583929,
      "device_id": 66306,
      "size": 1103,
      "links": 1,
      "content": [
        "1da0b6258ab486c9eeff33a2f5b75dd9b2429fccceb476e3cc800a8ad05b7240"
      ]
    },
    {
      "name": "upgrade-disable-web.config.php",
      "type": "file",
      "mode": 420,
      "mtime": "2023-12-27T10:48:10.424799432+01:00",
      "atime": "2023-12-27T10:48:10.424799432+01:00",
      "ctime": "2023-12-27T19:05:28.408866058+01:00",
      "uid": 33,
      "gid": 33,
      "inode": 9583930,
      "device_id": 66306,
      "size": 60,
      "links": 1,
      "content": [
        "282e2f702823e7fc4315a58b76644dcafed71ea4faad0c9af4f9a031e8d54ac2"
      ]
    }
  ]
}
```
UIDs and GIDs here seems fine to me... maybe you could first try to see if you can replicate the issue?

**Reply by aawsome** [reply]
Yes, in fact they seem to be fine. I'll try to reproduce some examples with non-existing uids.

**Reply by aawsome** [reply]
@LeonardoMantovani I finally found the time to try to reproduce your behavior - unfortunately, I couldn't. 
I have a file `info.txt` in my backup with UID/GID 1004 - which is not present in this system:
```
rustic ls -l --numeric-uid-gid <SNAP>:/<PATH>/info.txt
   1004     1004       129 15 Nov 20:45 "info.txt" 
```

after I run `rustic restore --numeric-id <SNAP>:/<PATH>/info.txt test/`, I see this file:
```
ls -al test/
insgesamt 12
drwxrwxr-x  2 alex-dev alex-dev 4096 Feb  2 06:32 .
drwxr-x--- 19 alex-dev alex-dev 4096 Feb  2 06:32 ..
-rw-r--r--  1     1004     1004  129 Nov 15  2016 info.txt
```

Note that I ran the restore command with `root`.

Running it with user permissions gave the `[WARN] restore "info.txt": setting UID/GID failed.` and the user/group was not set. But in this case, `chown` also returned an error and wasn't able to set ownership.

Can you please re-test if there really is a problem restoring non-existing UID/GID? Can you try the latest rustic version or the nightly version?

**Reply by aawsome** [reply]
@LeonardoMantovani Can you please also provide information about the environment you are using? Could be something OS- or platform-specific, too...

**Reply by LeonardoMantovani** [reply]
Thank you for taking the time to investigate on this, I really appreciate it.
I'm running NixOS 23.11 with rustic 0.6.1.

The issue is probably that, since I was executing the backups as the normal user, I didn't thought that it would have been required to be root user to execute the restore.
But now I tried and you're right, if the restore command is done by the root user permissions are restored correctly.
And this makes sense, since as you were saying the *chown* command returns an error if executed as standard user and the *restic restore* command behaves exactly the same way.

My bad that I didn't notice it before, and sorry for bothering you with that!

**Reply by aawsome** [reply]
no worries - better to get a bug report which turns out to be no bug than to not get a bug report of a real problem :-)

**Reply by simonsan** [reply]
I opened https://github.com/rustic-rs/docs/issues/58 to clarify that in the docs for the future. 👍🏽

**Comment by aawsome** [comment, marked as answer]
The option `--no-ownership` is now available in the latest beta builds.

**Comment by aawsome** [comment]
> (Unfortunately there is so far no `rustic ls --numeric-ic` or something like this)

I opened #1019 to add `--numeric-uid-gid` to `rustic ls`.

**Reply by LeonardoMantovani** [reply]
Hey, thank you for addressing the issue that quickly!
So now from the next release (I guess rustic 0.6.2) I should be able to restore non-existing UIDs ang GIDs properly by using `rustic -r <my_repo> restore latest . --numeric-uid-gid`, right?
And, if so, any idea about approximately when this next release will be available?

**Reply by simonsan** [reply]
The next release will be a release containing breaking changes, so it will be 0.7.0. And we can't give any estimations on upcoming release dates, as we are doing that all in our spare time. But it shouldn't be too far away. It's most likely happening in February, but no promises. :)

**Reply by LeonardoMantovani** [reply]
Thank you! And don't worry, I totally understand, I was asking just because of curiosity.
Good luck with 0.7.0 and thank you for all your work on this project

**Comment by aawsome** [comment]
> So now from the next release (I guess rustic 0.6.2) I should be able to restore non-existing UIDs ang GIDs properly by using `rustic -r <my_repo> restore latest . --numeric-uid-gid`, right?

No. So far I only added the possibility to examine the uid/gid using the `rustic ls` command.
I wasn't able to work on reproducing your issue so far, unfortunately. My priorities are currently on the upcoming release.
